numba/cgutils.py

```diff
@@ -676,6 +676,16 @@ def get_item_pointer2(builder, data, shape, strides, layout, inds,
             indices.append(selected)
     else:
         indices = inds
+    if boundscheck:
+        for ind, dimlen in zip(indices, shape):
+            msg = "index is out of bounds"
+            out_of_bounds_upper = builder.icmp_signed('>=', ind, dimlen)
+            with if_unlikely(builder, out_of_bounds_upper):
+                context.call_conv.return_user_exc(builder, IndexError, (msg,))
+            out_of_bounds_lower = builder.icmp_signed('<', ind, ind.type(0))
+            with if_unlikely(builder, out_of_bounds_lower):
```

Comment:
Should I merge these two ifs? I didn't see how to do a boolean or in the llvmlite documentation.

Reply:
You possibly want something like this:
```python
            out_of_bounds_upper = builder.icmp_signed('>=', ind, dimlen)
            out_of_bounds_lower = builder.icmp_signed('<', ind, ind.type(0))
            predicate = builder.or_(out_of_bounds_upper, out_of_bounds_lower)
            with if_unlikely(builder, predicate):
                context.call_conv.return_user_exc(builder, IndexError, (msg,))
```
Docs: http://llvmlite.pydata.org/en/latest/user-guide/ir/ir-builder.html#llvmlite.ir.IRBuilder.or_

Reply:
It says `or_` is bitwise OR (like `|` instead of `||`). I don't know if that matters here. I mainly want to make sure I generate code that LLVM optimizes correctly (how can I check that?).

Reply:
It is indeed bitwise `or`, suggested that for same net effect, though it will likely generate an `or` op whereas a logical `or` would likely generate jump(s). To get something logical `or`-like (i.e. `||`) it's is a bit like what you have already but the branches need nesting so that the second condition is only checked if the first passes, I'm assuming that's what you are thinking of?

You can check the LLVM IR with `NUMBA_DUMP_LLVM=1` as an env var, or the `.inspect_llvm` method on the dispatcher. The assembly can be seen with the `.inspect_asm` method on the dispatcher. How were you hoping it to optimise?

Reply:
I mean if the index is already known to be less than the dimension or nonnegative it should optimize out the `if`.

Reply:
I'm not sure how you'd do that at compile time? It sound like you want to compile a version of this code based on the run time values?

Reply:
If I'm reading the LLVM correctly (and I'm not very good at reading LLVM, so I could be wrong), it saves the IndexError as a predefined constant. So that adds a bit of overhead to every function call (though it looks like it does it for every exception, so maybe that's not my problem to solve here). Also it implies that it would be more efficient to have only one `return_user_exc` here.

Reply:
> I'm not sure how you'd do that at compile time? It sound like you want to compile a version of this code based on the run time values?

I mean if you have `x[n]` where `n` is defined in such a way that it is statically impossible for it to be negative, then it does not need to have the second `if`. I don't know how smart LLVM is, but presumably this can happen in the optimization passes.

It's possible I'm thinking about this too much, but since this does incur a performance overhead, I want to make sure that overhead is as small as possible.

Reply:
I think under very specific conditions it might be possible to do this, but all the information would likely have to be constant at compile time. Also, *any* `if` from the pattern being added here will likely create a branch in a tight indexing loop and prevent vectorization.

Reply:
> If I'm reading the LLVM correctly (and I'm not very good at reading LLVM, so I could be wrong), it saves the IndexError as a predefined constant. So that adds a bit of overhead to every function call (though it looks like it does it for every exception, so maybe that's not my problem to solve here). Also it implies that it would be more efficient to have only one `return_user_exc` here.

I'm not sure how this adds overhead to a function call, any chance you could please expand on this? Thanks.

Reply:
> I think under very specific conditions it might be possible to do this, but all the information would likely have to be constant at compile time. Also, any if from the pattern being added here will likely create a branch in a tight indexing loop and prevent vectorization.

So if I have something like

```py
A = 0
for i in range(x.shape[0]):
    A += x[i]
```

it won't be able to deduce that `0 <= i < x.shape[0]` and so the bounds checking can be optimized away?

> I'm not sure how this adds overhead to a function call, any chance you could please expand on this? Thanks.

I think I misread the LLVM. The exceptions are saved as constant pickles on the functions, but are only unpickled when needed. So it adds a memory overhead to the function definition, but that's it. It's probably better to not have the exception there twice but maybe not a huge deal.

More importantly, since the exceptions are precomputed, I don't see how to raise an exception with a dynamic message, namely including the dimension in the error message like NumPy does. Would we need to add support for pickling a function that returns the exception with the given message? Or is this already something that can be done? If it's not easy to do this, I might skip it for this PR, if that's alright.

Reply:
> > I think under very specific conditions it might be possible to do this, but all the information would likely have to be constant at compile time. Also, any if from the pattern being added here will likely create a branch in a tight indexing loop and prevent vectorization.
> 
> So if I have something like
> 
> ```python
> A = 0
> for i in range(x.shape[0]):
>     A += x[i]
> ```
> 
> it won't be able to deduce that `0 <= i < x.shape[0]` and so the bounds checking can be optimized away?

In theory, yes, but it depends on what code gets generated. I think that if an unsigned comparison operator were used in the bounds check then LLVM is more likely to be able to work it out, if a signed comparison is done then it won't. I just generated the assembly for:

  1. no boundscheck
  2. boundscheck with signed comparison
  3. boundscheck with unsigned comparison


Cases 1. and 3.are reasonably similar but 1. has a deeper loop unroll. 2. has no vectorisation because there's a test/jump in the middle of the loop.

I think more can be done in terms of identifying unsigned ranges and suspect that this is likely to help.

> 
> > I'm not sure how this adds overhead to a function call, any chance you could please expand on this? Thanks.
> 
> I think I misread the LLVM. The exceptions are saved as constant pickles on the functions, but are only unpickled when needed. So it adds a memory overhead to the function definition, but that's it. It's probably better to not have the exception there twice but maybe not a huge deal.
> 
> More importantly, since the exceptions are precomputed, I don't see how to raise an exception with a dynamic message, namely including the dimension in the error message like NumPy does. Would we need to add support for pickling a function that returns the exception with the given message? Or is this already something that can be done? If it's not easy to do this, I might skip it for this PR, if that's alright.

Numba stores exception strings as compile time constants so you cannot raise an exception with a dynamically generated string. So yes, skipping it is fine :)

Reply:
I made all `range(x)` calls unsigned in https://github.com/numba/numba/pull/4446 and was reminded of the amount of work needed to implement this correctly due to the NumPy casting rule (generalizing as):
`uint* <op> int* -> float`.

Reply:
> Numba stores exception strings as compile time constants so you cannot raise an exception with a dynamically generated string. So yes, skipping it is fine :)

I think you could serialize a function that creates the IndexError rather than the IndexError itself. Then modify `return_user_exc` so it can call the function to create the exception before raising it. 

The thing I'm not sure about are the builder instructions for converting the index, shape, and axis into strings so they can be put in the message.